src/commonMain/kotlin/fr/acinq/lightning/channel/InteractiveTx.kt

```diff
@@ -410,45 +420,42 @@ data class SharedTransaction(
     fun sign(session: InteractiveTxSession, keyManager: KeyManager, fundingParams: InteractiveTxParams, localParams: LocalParams, remoteNodeId: PublicKey): PartiallySignedSharedTransaction {
         val unsignedTx = buildUnsignedTx()
         val sharedSig = fundingParams.sharedInput?.sign(keyManager.channelKeys(localParams.fundingKeyPath), unsignedTx)
-        val sharedOutput = fundingParams.sharedInput?.let { i -> mapOf(i.info.outPoint to i.info.txOut) } ?: mapOf()
-        val localOutputs = localInputs.associate { i -> i.outPoint to i.txOut }
-        val remoteOutputs = remoteInputs.associate { i -> i.outPoint to i.txOut }
-        val previousOutputsMap = sharedOutput + localOutputs + remoteOutputs
-        val previousOutputs = unsignedTx.txIn.map { previousOutputsMap[it.outPoint]!! }.toList()
+        // NB: the order in this list must match the order of the transaction's inputs.
+        val previousOutputs = unsignedTx.txIn.map { spentOutputs[it.outPoint]!! }
 
-        // nonces that we've received for all musig2 swap-in
-        val receivedNonces: Map<Long, IndividualNonce> = when (session.txCompleteReceived) {
+        // Public nonces for all the musig2 swap-in inputs (local and remote).
+        // We have verified that one nonce was provided for each input when receiving `tx_complete`.
+        val remoteNonces: Map<Long, IndividualNonce> = when (session.txCompleteReceived) {
             null -> mapOf()
             else -> (localInputs.filterIsInstance<InteractiveTxInput.LocalSwapIn>() + remoteInputs.filterIsInstance<InteractiveTxInput.RemoteSwapIn>())
                 .sortedBy { it.serialId }
                 .zip(session.txCompleteReceived.publicNonces)
                 .associate { it.first.serialId to it.second }
         }
 
-
         // If we are swapping funds in, we provide our partial signatures to the corresponding inputs.
-        val swapUserSigs = unsignedTx.txIn.mapIndexed { i, txIn ->
+        val legacySwapUserSigs = unsignedTx.txIn.mapIndexed { i, txIn ->
             localInputs
                 .filterIsInstance<InteractiveTxInput.LocalLegacySwapIn>()
                 .find { txIn.outPoint == it.outPoint }
                 ?.let { input -> keyManager.swapInOnChainWallet.signSwapInputUserLegacy(unsignedTx, i, input.previousTx.txOut) }
         }.filterNotNull()
-
         val swapUserPartialSigs = unsignedTx.txIn.mapIndexed { i, txIn ->
             localInputs
                 .filterIsInstance<InteractiveTxInput.LocalSwapIn>()
-                .find { txIn.outPoint == it.outPoint && session.secretNonces.containsKey(it.serialId) && receivedNonces.containsKey(it.serialId) }
+                .find { txIn.outPoint == it.outPoint }
```

Comment:
I don't think it can? On the contrary, if it throws an NPE, that means there is a bug in the code. We create nonces for each `serial_id` when sending/receiving `tx_add_input`, so we're guaranteed that they can be found in the map, aren't we?

Reply:
For our `secretNonces` yes, but not for received nonces (we just check the number of received nonces, not their serial ids).

Reply:
Done in https://github.com/ACINQ/lightning-kmp/pull/592/commits/47df69e3f1dd70b51a25aac51f22a2eb415d211e

Reply:
Ok I see that now: if we're missing a nonce it means that there is indeed a bug in our code (though the additional check is harmless).